yellow-paper/docs/gas-and-fees/gas-metering-and-fees.md

```diff
@@ -0,0 +1,409 @@
+- `inclusion_fee`
+
+### CallContext
+
+- `gas_limits`
+- `transaction_fee`
+  - Only non-zero during teardown
+
+### PrivateKernelPublicInputs
+
+- `gas_limits`
+- `end`: `PrivateAccumulatedData`
+
+### PrivateAccumulatedData
+
+- `gas_used`
+
+### PublicKernelPublicInputs
+
+- `gas_limits`
+- `end`: `PublicAccumulatedData`
+- `end_non_revertible`: `PublicAccumulatedData`
+
+### PublicAccumulatedData
+
+- `gas_used`
+
+### PublicCircuitPublicInputs
+
+- `fee_paid`
+  - false unless the `pay_fee` function was called in teardown
+
+### KernelCircuitPublicInputs
+
+- `gas_limits`
+- `transaction_fee`
+  - Summed during public kernel tail
+  - Zero-ed during public kernel teardown if `fee_paid` is false
+- `end`: `AccumulatedData`
+
+### AccumulatedData
+
+- `gas_used`
+
+### GlobalVariables
+
+- `fee_per_da_gas`
+- `fee_per_compute_gas`
+
+
+# Pure private transactions
+
+Purely private transactions cannot pay a fee, and thus the `transaction_fee` is set to 0.
+
+However, they still consume gas, and thus the `da_gas_used` and `compute_gas_used` must be tracked.
+
+```
+kernel_public_inputs.gas_limits := private_kernel_public_inputs.gas_limits
+
+kernel_public_inputs.end.gas_used.da_gas_used :=
+    FIXED_DA_GAS
+    + private_kernel_public_inputs.end.gas_used.da_gas_used
+
+kernel_public_inputs.end.compute_gas_used :=
+    private_kernel_public_inputs.end.compute_gas_used
+
+# make assertions that gas limits are not exceeded
+
+# purely private transactions cannot pay any fee
+kernel_public_inputs.transaction_fee := 0
+```
+
+# Transaction with public component
+
+If a transaction has a public component, `end_non_revertible.gas_used.da_gas_used` is initialized to `FIXED_DA_GAS` in the `PrivateKernelTailToPublic` kernel.
+
+### Interaction with the public VM
+
+The public VM retains the current amount of `daGasLeft` in its [MachineState](../public-vm/state.md#machine-state).
+
+This amount must be initialized for each public kernel iteration as:
+
+```
+daGasLeft := public_kernel_public_inputs.da_gas_limit 
+             - public_kernel_public_inputs.end_non_revertible.da_gas_used
+             - public_kernel_public_inputs.end.da_gas_used
+```
+
+At the end of execution, the public VM will have an updated `daGasLeft` value.
+
+This value will be used by the public kernel to update the `da_gas_used` in the `PublicAccumulatedData` struct corresponding to the current phase:
+
+```
+# if in setup 
+public_kernel_public_inputs.end_non_revertible.gas_used.da_gas_used := 
+        public_kernel_public_inputs.gas_limits.da_gas_limit
+          - daGasLeft
+          - public_kernel_public_inputs.end.gas_used.da_gas_used
+
+# if in app logic
+public_kernel_public_inputs.end.gas_used.da_gas_used := 
+        public_kernel_public_inputs.gas_limits.da_gas_limit
+          - daGasLeft
+          - public_kernel_public_inputs.end_non_revertible.gas_used.da_gas_used
+```
+
+
+It will be the same for the `compute_gas_used`.
+
+### Out of gas
```

Comment:
Can we add a section on what happens during a revert in a public function? I'd expect that the compute gas consumed so far is left as-is, but the DA gas for revertible is cleared out, since reverted side-effects don't go on-chain, right?